File: backend/app/schemas/note.py
"""
Pydantic schemas for note-related operations.
"""

from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class NoteBase(BaseModel):
    """Base schema for note data."""
    content: str = Field(..., min_length=1, max_length=5000, description="Note content")
    timestamp: Optional[int] = Field(None, ge=0, description="Video timestamp in seconds")


class NoteCreate(NoteBase):
    """Schema for creating a new note."""
    lecture_id: int = Field(..., gt=0, description="ID of the lecture")


class NoteUpdate(BaseModel):
    """Schema for updating an existing note."""
    content: Optional[str] = Field(None, min_length=1, max_length=5000, description="Note content")
    timestamp: Optional[int] = Field(None, ge=0, description="Video timestamp in seconds")


class NoteResponse(NoteBase):
    """Schema for note response data."""
    id: int
    user_id: int
    lecture_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NoteListResponse(BaseModel):
    """Schema for paginated note list response."""
    notes: list[NoteResponse]
    total: int
    page: int
    per_page: int
    total_pages: int
    has_next: bool
    has_prev: bool


class NoteSearchFilters(BaseModel):
    """Schema for note search and filtering."""
    lecture_id: Optional[int] = None
    course_id: Optional[int] = None
    search: Optional[str] = Field(None, max_length=200, description="Search in note content")
    page: int = Field(1, ge=1, description="Page number")
    per_page: int = Field(20, ge=1, le=100, description="Items per page")

File: frontend/app/types/note.ts
/**
 * Types for note-taking functionality
 */

export interface Note {
    id: number;
    user_id: number;
    lecture_id: number;
    content: string;
    timestamp?: number; // Video timestamp in seconds
    created_at: string;
    updated_at: string;
}

export interface NoteCreate {
    lecture_id: number;
    content: string;
    timestamp?: number;
}

export interface NoteUpdate {
    content?: string;
    timestamp?: number;
}

export interface NoteListResponse {
    notes: Note[];
    total: number;
    page: number;
    per_page: number;
    total_pages: number;
    has_next: boolean;
    has_prev: boolean;
}

export interface NoteSearchFilters {
    lecture_id?: number;
    course_id?: number;
    search?: string;
    page?: number;
    per_page?: number;
}